apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-common/src/main/java/co/elastic/apm/agent/es/restclient/ElasticsearchRestClientInstrumentationHelperImpl.java

```diff
@@ -0,0 +1,145 @@
+import org.elasticsearch.client.ResponseException;
+import org.elasticsearch.client.ResponseListener;
+import org.jctools.queues.atomic.AtomicQueueFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+import static org.jctools.queues.spec.ConcurrentQueueSpec.createBoundedMpmc;
+
+public class ElasticsearchRestClientInstrumentationHelperImpl implements ElasticsearchRestClientInstrumentationHelper<HttpEntity, Response, ResponseListener> {
+
+    public static final String SEARCH_QUERY_PATH_SUFFIX = "_search";
+    public static final String SPAN_TYPE = "db";
+    public static final String ELASTICSEARCH = "elasticsearch";
+    public static final String SPAN_ACTION = "request";
+    private static final int MAX_POOLED_ELEMENTS = 256;
+    private final ElasticApmTracer tracer;
+
+    private final ObjectPool<ResponseListenerWrapper> responseListenerObjectPool;
+
+    public ElasticsearchRestClientInstrumentationHelperImpl(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        responseListenerObjectPool = QueueBasedObjectPool.ofRecyclable(
+            AtomicQueueFactory.<ResponseListenerWrapper>newQueue(createBoundedMpmc(MAX_POOLED_ELEMENTS)),
+            false,
+            new ResponseListenerAllocator());
+    }
+
+    private class ResponseListenerAllocator implements Allocator<ResponseListenerWrapper> {
+        @Override
+        public ResponseListenerWrapper createInstance() {
+            return new ResponseListenerWrapper(ElasticsearchRestClientInstrumentationHelperImpl.this);
+        }
+    }
+
+    @Override
+    @Nullable
+    public Span createClientSpan(String method, String endpoint, @Nullable HttpEntity httpEntity) {
+        final TraceContextHolder<?> activeSpan = tracer.getActive();
+        if (activeSpan == null || !activeSpan.isSampled()) {
+            return null;
+        }
+
+        // Don't record nested spans. In 5.x clients the instrumented sync method is calling the instrumented async method
+        if (activeSpan instanceof Span && ELASTICSEARCH.equals(((Span) activeSpan).getSubtype())) {
+            return null;
+        }
+
+        Span span = activeSpan.createSpan()
+            .withType(SPAN_TYPE)
+            .withSubtype(ELASTICSEARCH)
+            .withAction(SPAN_ACTION)
+            .appendToName("Elasticsearch: ").appendToName(method).appendToName(" ").appendToName(endpoint);
+        span.getContext().getDb().withType(ELASTICSEARCH);
+        span.activate();
+
+        if (span.isSampled()) {
+            span.getContext().getHttp().withMethod(method);
+            if (endpoint.endsWith(SEARCH_QUERY_PATH_SUFFIX)) {
+                if (httpEntity != null && httpEntity.isRepeatable()) {
+                    try {
+                        IOUtils.readUtf8Stream(httpEntity.getContent(), span.getContext().getDb().withStatementBuffer());
+                    } catch (IOException e) {
+                        // We can't log from here
```

Comment:
we actually can :)

Reply:
We sure can, noticed that and forgot to log 🤦‍♂️